### PR TITLE
fix: clean pending requests

### LINKED ISSUE #14: Test: on firefox

### LINKED ISSUE #187: bug: the last pendingRequest page is open if closed with "x" button.

### PR DESCRIPTION
**Pull request by iskdrews**
## Explanation

chore:
- Making sure to clean any pendingRequested related to any opened window which is closed from the user via clicking on "x" button.

enhance:
   - Adding a button for creating a new identity in the demo website.

## More Information

Fixes #187 
Closes #14
Related to #140  
Related to #183 

## Pre-Merge Checklist

- [x] PR template is filled out
- [ ] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added
- [x] PR is linked to the appropriate GitHub issue
- [x] PR has been added to the appropriate release Milestone

### + If there are functional changes:

- [ ] Manual testing complete & passed
- [ ] "Extension QA Board" label has been applied

> PR template source from [github.com/MetaMask](https://github.com/MetaMask)

### CODE REVIEW

File: demo/index.tsx [hunk 1]
```diff
@@ -147,6 +147,10 @@ function App() {
     setIdentityCommitment(idCommitment);
   }, [client, setIdentityCommitment]);
 
+  const createIdentity = useCallback(async () => {
```
Comment (0xmad) [inline demo/index.tsx:150]: I think you don't need async/await here and you can use onClick handler without arrow function.

File: demo/index.tsx [hunk 2]
```diff
@@ -233,6 +238,13 @@ function App() {
         <br />
       </div>
 
+      <hr />
+      <div>
+        <h2>Create a new Identity</h2>
+        <button onClick={() => createIdentity()}>Create</button> <br />
```
Comment (0xmad) [inline demo/index.tsx:244]:
```suggestion
        <button onClick={createIdentity}>Create</button> <br />
```

File: src/types/index.ts [hunk 3]
```diff
@@ -45,6 +45,7 @@ export enum PendingRequestType {
 
 export type PendingRequest = {
   id: string;
+  windowId: number | undefined;
```
Comment (0xmad) [inline src/types/index.ts:48]:
```suggestion
  windowId?: number;
```

File: src/background/zkKeeper.ts [hunk 4]
```diff
@@ -37,6 +37,22 @@ export default class ZkKeeperController extends Handler {
...
     this.walletService = new WalletService();
     this.lockService = LockService.getInstance();
+
+    BrowserUtils.addRemoveWindowListener((windowId: number) => {
+      log.debug("Inside removeWindow onRemove");
+
+      try {
+        log.debug("Inside removeWindow onRemove locked");
+        if (BrowserUtils.cached?.id === windowId) {
+          BrowserUtils.cached = null;
```
Comment (0xmad) [inline src/background/zkKeeper.ts:47]: I don't think it's a good idea to change cached value outside `BrowserUtils` class.
I'd move this logic inside the class.

File: src/background/controllers/browserUtils.ts [hunk 5]
```diff
@@ -22,24 +22,10 @@ interface OpenPopupArgs {
 }
 
 class BrowserUtils {
-  private cached: Windows.Window | null = null;
+  public cached: Windows.Window | null = null;
 
   public constructor() {
```
Comment (0xmad) [inline src/background/controllers/browserUtils.ts:27]: You can remove constructor if you don't need it

File: src/background/controllers/requestManager.ts [hunk 6]
```diff
@@ -30,31 +29,44 @@ export default class RequestManager extends EventEmitter2 {
     return true;
   };
 
-  public addToQueue = async (type: PendingRequestType, payload?: unknown): Promise<string> => {
+  public finilizeRequestOnRemovedWindow = async (windowId: number | undefined): Promise<boolean> => {
```
Comment (0xmad) [inline src/background/controllers/requestManager.ts:32]:
```suggestion
  public finilizeRequestOnRemovedWindow = async (windowId?: number): Promise<boolean> => {
```

File: src/background/controllers/requestManager.ts [hunk 7]
```diff
@@ -30,31 +29,44 @@ export default class RequestManager extends EventEmitter2 {
...
+      this.pendingRequests = this.pendingRequests.filter((pendingRequests) => pendingRequests.windowId !== windowId);
+      await pushMessage(setPendingRequest(this.pendingRequests));
+    }
+
+    return true;
+  };
+
+  public addToQueue = async (
+    type: PendingRequestType,
+    windowId: number | undefined,
```
Comment (0xmad) [inline src/background/controllers/requestManager.ts:43]:
```suggestion
    windowId?: number,
```

**Review by 0xmad — CHANGES_REQUESTED** [review]
@iskdrews thanks for the fix!
There are some comments related to encapsulation for BrowserUtils class.

**Comment by iskdrews** [comment]
Hi @0xmad thanks so much for your comments. 
I tried to move the code back to BrowserUtils class, but there is a cyclic dependency happened with `BrowserUtils` and `RequestManager` please review my workaround to resolve this. Thanks.

**Review by 0xmad — APPROVED** [review]
@iskdrews thanks! Feel free to merge when you are ready :)

**Pull request merged**